// cis/db.js
var mongodb = require('mongodb');
var fs          = require('fs');

var DB = {
    host: "192.168.113.11", // Server Prod
    // host: "192.168.110.152", // Server Prod (when connected to the VPN)
    // host: "localhost", 
    port: "27017", // "10712",
    user: "admin",
    password: "Password.01",
    useauthentication: false
}

var init = function(dbName) {
    var server = new mongodb.Server(DB.host, DB.port, {
        auto_reconnect: true
    });

    var dbConnect = new mongodb.Db(dbName, server, {safe: false});
    return dbConnect; // Returns the instance
};

var connectDB = function (mongoinstance, dbName, callback) {
    
    var server = new mongodb.Server(DB.host, DB.port, {
        auto_reconnect: true
    });

    if (dbName == false) {
        var dbConnect = new mongodb.Db(DB.db, server, {
            safe: false
        });
        dbConnect.open(function (err, db) {
            callback(err, db);
        });
    } else {
        // var dbConnect = mongoinstance; 
        var dbConnect = new mongodb.Db(dbName, server, {safe: false});

        if (!dbConnect.serverConfig.isConnected())
        {
            dbConnect.open(function (err, db) {
                if (!err)
                {
                    if (DB.useauthentication)
                    {
                        db.authenticate(DB.user, DB.password, function(err, res) {
                            if(!err) {
                                callback(err, db);
                            } else {
                                callback(err, db);
                            }
                        });
                    }
                    else
                    {
                        callback(err, db);
                    }
                }
                else
                {
                    callback(err, db);
                }
            });
        }
        else{
            // Already open
            callback(null, dbConnect);
        }
    }
};

exports.dbOps = function(){
    return{
        init: function (dbName) {
            return init(dbName);
        },
        connectDB : function (mongoinstance, dbName, callback) {
            connectDB(mongoinstance, dbName, callback);
        }
    }
}();
